### ISSUE TITLE
Log fetches for EcsRunTaskOperator return incomplete logs

### ISSUE THREAD

**Issue by yaningz** [apache/airflow#43717]
### Apache Airflow Provider(s)

amazon

### Versions of Apache Airflow Providers

apache-airflow==2.10.1
apache-airflow-providers-airbyte==3.9.0
apache-airflow-providers-amazon==8.28.0
apache-airflow-providers-databricks==6.9.0
apache-airflow-providers-docker==3.13.0
apache-airflow-providers-slack==8.9.0
apache-airflow-providers-snowflake==5.7.0

### Apache Airflow version

2.10.1

### Operating System

Amazon Linux

### Deployment

Amazon (AWS) MWAA

### Deployment details

Our MWAA environments are managed through Terraform. We install 1 provider outside of those available in the public providers, as well as 1 internal provider.

### What happened

Related to https://github.com/apache/airflow/issues/40875

We previously reported the same issue, and after upgrading to a newer version of Airflow and the AWS provider that included the change, we were still able to detect an issue with logs appearing across log groups for ECS tasks.

In all versions of Airflow between 2.7 and 2.10, we have observed logs missing from the `*-Task` log group of our MWAA deployment. As in https://github.com/apache/airflow/issues/40875, we continue to use the `awslogs_group` argument for the EcsRunTaskOperator to specify the logs group that the UI should pull from. 

### **On successful task execution**

The log group associated with our ECS cluster ( `/ecs/airflow2/airflow2` ) retains all logs, but these are not fully reflected in the `*-Task` log group ( `airflow-airflow2-Task` ). The following files demonstrate a pair of log streams that should contain all the same events, but all lines after `[0m13:11:27  7 of 30 START test not_null_disease_location_proximity_context_results_parsed_updt_dttm  [RUN]` are missing in the log stream for the `*-Task` log group.

ECS cluster logstream (log group `/ecs/airflow2/airflow2` ): [202411015-ecs:airflow2:airflow-logstream.csv](https://github.com/user-attachments/files/17637213/202411015-ecs.airflow2.airflow-logstream.csv)

MWAA logstream (log group  `airflow-airflow2-Task` ): [20241105-airflow-airflow2-Task-logstream.csv](https://github.com/user-attachments/files/17637214/20241105-airflow-airflow2-Task-logstream.csv)

### **On unsuccessful task execution**

Additionally, we found that `::group::Post task execution logs` were able to retrieve remaining logs during failed tasks, but successful tasks left the remaining logs behind. The following is an example of a failed task that includes all log events as part of post task execution logs pulled into the log stream.

ECS cluster logstream (log group `/ecs/airflow2/airflow2` ): [20241008-ecs:airflow2:airflow2-logstream.csv](https://github.com/user-attachments/files/17637328/20241008-ecs.airflow2.airflow2-logstream.csv)

MWAA logstream (log group  `airflow-airflow2-Task` ): [20241008-airflow-airflow-Task-logstream.csv](https://github.com/user-attachments/files/17637329/20241008-airflow-airflow-Task-logstream.csv)

### **EcsRunTaskOperator call**

This is the most current version of our call to the EcsRunTaskOperator class:

```
        super().__init__(
            task_definition=task_definition,
            cluster=airflow_env_name,
            overrides={
                "containerOverrides": [
                    {
                        "name": "dbt-om1-task",
                        "command": command_list,
                    },
                ],
            },
            launch_type="FARGATE",
            network_configuration={
                "awsvpcConfiguration": {
                    "subnets": [
                        os.environ.get("AIRFLOW__VAR__PRIMARY_SUBNET_ID"),
                        os.environ.get("AIRFLOW__VAR__SECONDARY_SUBNET_ID"),
                        os.environ.get( "AIRFLOW__VAR__TERTIARY_SUBNET_ID"),
                    ],
                    "securityGroups": [
                        os.environ.get("AIRFLOW__VAR__SECURITY_GROUP_ID")
                    ],
                    "assignPublicIp": "DISABLED",
                },
            },
            awslogs_group=f"/ecs/airflow2/airflow2",
            awslogs_region="us-east-1",
            awslogs_stream_prefix=f"airflow2/dbt-om1-task",
            awslogs_fetch_interval=timedelta(seconds=30),
            propagate_tags="TASK_DEFINITION",
            **kwargs,
        )
```

### What you think should happen instead

All log events in the ECS log group should be pulled forward by the MWAA `-*Task` log group using the `task_log_fetcher`.

### How to reproduce

As in https://github.com/apache/airflow/issues/40875, we are using a large (~2GB) custom DBT image inside our taskdef. This has happened with every execution of the EcsRunTaskOperator against our custom DBT image.

After triggering a task using the EcsRunTaskOperator, all logs should be copied forward from the log stream in the ECS cluster's log group to the log stream within the MWAA task log group.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by mjordyn** [comment]
I encountered a very similar issue in Airflow 2.10.1 with apache-airflow-providers-amazon==8.28.0. However, in my case, I was running Airflow in a Docker container (Docker Compose), not using MWAA, and the logs missing from the Airflow UI were the first few from each of my ECS tasks. These logs were present in the ECS task CloudWatch log streams. Adding a 10 second sleep statement before the logger.info calls as a workaround caused those logs to appear as expected in the Airflow UI. All of this held true for a few separate `EcsRunTaskOperator` steps calling different ECS tasks.

**Comment by o-nikolas** [comment]
@vincbeck Looks like the issue is still happening even after #41515 :/

**Comment by vincbeck** [comment]
> [@vincbeck](https://github.com/vincbeck) Looks like the issue is still happening even after [#41515](https://github.com/apache/airflow/pull/41515) :/

😢

**Comment by o-nikolas** [comment]
Hello @yaningz, I'm just looking into this again. You mention:

>As in https://github.com/apache/airflow/issues/40875, we are using a large (~2GB) custom DBT image inside our taskdef. This has happened with every execution of the EcsRunTaskOperator against our custom DBT image.

Is the issue only occurring with this image? Do you use other images which do not see this behaviour?

**Comment by o-nikolas** [comment]
Hello @mjordyn 
Regarding this:
> Adding a 10 second sleep statement before the logger.info calls as a workaround caused those logs to appear as expected in the Airflow UI. All of this held true for a few separate EcsRunTaskOperator steps calling different ECS tasks.

1) Before which logger.info calls did you add sleeps? Was it [here](https://github.com/apache/airflow/blob/b275e79654f54b6373c392d4750811d6373c40d9/providers/amazon/src/airflow/providers/amazon/aws/utils/task_log_fetcher.py#L75)?
2) Did you do any testing regarding how long of a sleep was required? Was 10s the minimum?
3) Was this an experiment only, or did you leave this change in place? I.e. were you happy to wait longer for full logs?

**Comment by yaningz** [comment]
Hello @o-nikolas, regarding this question:

> Is the issue only occurring with this image? Do you use other images which do not see this behaviour? Also, is it always the tail end of the logs that is missing? Or is it ever logs at the beginning or middle of the stream?

I need to refresh my memory on how this impacts ECS tasks other than with our image executing DBT, but the logs in the original issue are always truncated, never missing from the middle or beginning of the task. 

My belief is that this is caused by logs being pushed to the same timestamp in the logstream, and subsequently not being pulled because the timestamp is misaligned with the actual execution time. This should be visible in the logstreams attached for the successful task execution.

**Comment by mjordyn** [comment]
Hi @o-nikolas , thank you for taking a look at this.

> Before which logger.info calls did you add sleeps? Was it here?

Sorry, to clarify: I added the sleep to the beginning of my ECS tasks, before any of the `logger.info` calls in the task, to confirm that the info logs that were not showing up in the Airflow UI were in fact not showing up because of how early they were in the task, rather than due to some other confounding factor.

> Did you do any testing regarding how long of a sleep was required? Was 10s the minimum?

I did not -- I was just trying to confirm what I described above, and when I saw that a 10 second sleep caused all logs to appear, I took that as sufficient evidence of a correlation between a message being logged early in the task, and that message not showing up in the Airflow UI, and did not experiment further.

> Was this an experiment only, or did you leave this change in place? I.e. were you happy to wait longer for full logs?

This was just an experiment; I reverted the sleep addition after confirming the connection between logs being early on in the task, and not appearing in the UI.

**Comment by o-nikolas** [comment]
> Hi [@o-nikolas](https://github.com/o-nikolas) , thank you for taking a look at this.
> 
> > Before which logger.info calls did you add sleeps? Was it here?
> 
> Sorry, to clarify: I added the sleep to the beginning of my ECS tasks, before any of the `logger.info` calls in the task, to confirm that the info logs that were not showing up in the Airflow UI were in fact not showing up because of how early they were in the task, rather than due to some other confounding factor.
> 
> > Did you do any testing regarding how long of a sleep was required? Was 10s the minimum?
> 
> I did not -- I was just trying to confirm what I described above, and when I saw that a 10 second sleep caused all logs to appear, I took that as sufficient evidence of a correlation between a message being logged early in the task, and that message not showing up in the Airflow UI, and did not experiment further.
> 
> > Was this an experiment only, or did you leave this change in place? I.e. were you happy to wait longer for full logs?
> 
> This was just an experiment; I reverted the sleep addition after confirming the connection between logs being early on in the task, and not appearing in the UI.


Thanks for the info @mjordyn! Can you share a snippet of where you're putting the sleep in your ECS task? It sounds like you're putting it at the beginning of your task, then your task does it's actual work, then you see the logs? Or did you replace the actual work with only a sleep? If you have an actual code snipped (or if you could reproduce what change you made) and share it, that would be tremendous :)

**Comment by o-nikolas** [comment]
I've tried to reproduce the issue today, running ECS tasks that log 5, 50, 500, and 5000 lines of logging. Trying to run each one multiple times, and every time I got every log line reliably. So there is some piece missing either in my repro or the setup you all have running. This actually isn't too surprising, since if it was very easy to reproduce, we'd have many people cutting tickets (ECS is one of the most used AWS operators by far).

Is there anything in your infra setup that would cause cloudwatch to take longer to become consistent? Are you logging across regions? Or connecting through VCPE or some other gateway that would slow down network traffic out of your VPC? Any more specific you can provide will be very helpful to try reproduce something!

**Comment by yaningz** [comment]
Hi @o-nikolas, I've taken a look at our other tasks and logs. 

> I need to refresh my memory on how this impacts ECS tasks other than with our image executing DBT, but the logs in the original issue are always truncated, never missing from the middle or beginning of the task.

We are seeing this log truncation against other images. For proprietary reasons, I can't share the logs from the other task, but a simplified version of this other operator call is below. We do use similar patterns for our calls into the ECSRunTaskOperator regardless of the image, so it may be something in our configuration or infrastructure setup. 

```
class OtherCustomOperator(EcsRunTaskOperator):
    def __init__(
        self,
        image_tag: str,
        command: str,
        aws_account_name: str,
        airflow_env_name: str = DEFAULT_AIRFLOW_ENV_NAME,
        aws_conn_id: Optional[str] = "aws_default",
        **kwargs: Any,
    ) -> None:
        command_list = command.split()

        super().__init__(
            task_definition=ECS_NAME,
            cluster=airflow_env_name,
            overrides={
                "containerOverrides": [
                    {
                        "name": "other-custom-task",
                        "command": command_list,
                        "environment": [
                            {
                                "name": "AIRFLOW__VAR__AWS_ACCOUNT_NAME",
                                "value": aws_account_name,
                            }
                        ],
                    },
                ],
            },
            tags={
                "Local": str(is_local_airflow_instance())
            },
            launch_type="FARGATE",
            network_configuration={
                "awsvpcConfiguration": {
                    "subnets": [
                        os.environ.get("AIRFLOW__VAR__PRIMARY_SUBNET_ID"),
                        os.environ.get("AIRFLOW__VAR__SECONDARY_SUBNET_ID"),
                        os.environ.get("AIRFLOW__VAR__TERTIARY_SUBNET_ID"),
                    ],
                    "securityGroups": [
                        os.environ.get("AIRFLOW__VAR__SECURITY_GROUP_ID")
                    ],
                    "assignPublicIp": "DISABLED",
                },
            },
            awslogs_group=f"/ecs/{airflow_env_name}/{airflow_env_name}",
            awslogs_region="us-east-1",
            awslogs_stream_prefix=f"{airflow_env_name}/other-custom-task",
            awslogs_fetch_interval=timedelta(seconds=30),
            propagate_tags="TASK_DEFINITION",
            **kwargs,
        )

```

For our infrastructure, our MWAA environment is managed by Terraform, as is an ECS cluster specific to that MWAA environment. The `awslogs_group` that we pass into the operator call is a log group created separate from the ECS cluster, specific to task output logs. The target log group is specified within our taskdef (not configured on the cluster directly). 

I don't know of anything specific to our VPC setup that could be causing this or should be adding delay. I would have to ask one of our infrastructure engineers to help clarify if you have specific concerns about our networking that could be a cause.

Please let me know if there is any more information I can provide about our setup that may help in debugging.

**Comment by o-nikolas** [comment]
Thanks @yaningz

> For our infrastructure, our MWAA environment is managed by Terraform

Which version of MWAA are you using? Are you installing any other dependencies? If it all possible, can you share your requirements.txt? That way I can setup a similar environment to yours.

**Comment by yaningz** [comment]
@o-nikolas here's our requirements.txt (including 1 custom internal provider). We are still using `apache-airflow = "2.10.1"`, but have observed this log issue since an upgrade from 2.5 -> 2.7 over a year ago (but it wasn't discovered for several months).

```
--constraint "/usr/local/airflow/dags/constraints.txt"

airflow-provider-hex==0.1.9
airflow-provider-om1==2.10.0
apache-airflow-providers-airbyte==3.9.0
apache-airflow-providers-amazon==8.28.0
apache-airflow-providers-databricks==6.9.0
apache-airflow-providers-docker==3.13.0
apache-airflow-providers-slack==8.9.0
apache-airflow-providers-snowflake==5.7.0
opencv-python==4.10.0.84
pillow==11.0.0
```

I can't share our internal provider, but here are the specific dependencies of that package. There's a separate internal package embedded, but we began observing the log issues prior to that package being introduced to our Airflow environments.

```
[tool.poetry.dependencies]
python = ">=3.11,<3.12"
apache-airflow = "2.10.1"
apache-airflow-providers-amazon = "8.28.0"
apache-airflow-providers-databricks = "6.9.0"
apache-airflow-providers-docker = "3.13.0"
apache-airflow-providers-snowflake = {extras = ["common-compat"], version = "5.7.0"}
boto3 = "1.35.7"
pendulum = "^3"
boto3-stubs = {extras = ["ecs", "resourcegroupstaggingapi"], version = "1.35.7"}
<custom-internal-package> = {version = "^0.2.43", source = "codeartifact"}
deepmerge = ">=2.0"
```

Please let me know if there's anything else I can provide. Thank you for all the help!

**Comment by o-nikolas** [comment]
Thanks @yaningz, more questions in-line: 

> [@o-nikolas](https://github.com/o-nikolas) here's our requirements.txt (including 1 custom internal provider). We are still using `apache-airflow = "2.10.1"`, but have observed this log issue since an upgrade from 2.5 -> 2.7 over a year ago (but it wasn't discovered for several months).

Are you using MWAA version 2.10.1?

> 
> ```
> --constraint "/usr/local/airflow/dags/constraints.txt"
> 
> airflow-provider-hex==0.1.9
> airflow-provider-om1==2.10.0
> apache-airflow-providers-airbyte==3.9.0
> apache-airflow-providers-amazon==8.28.0
> apache-airflow-providers-databricks==6.9.0
> apache-airflow-providers-docker==3.13.0
> apache-airflow-providers-slack==8.9.0
> apache-airflow-providers-snowflake==5.7.0
> opencv-python==4.10.0.84
> pillow==11.0.0
> ```

So `airflow-provider-om1` is your package that is installed here that has the below requirements?

> 
> I can't share our internal provider, but here are the specific dependencies of that package. There's a separate internal package embedded, but we began observing the log issues prior to that package being introduced to our Airflow environments.
> 
> ```
> [tool.poetry.dependencies]
> python = ">=3.11,<3.12"
> apache-airflow = "2.10.1"
> apache-airflow-providers-amazon = "8.28.0"
> apache-airflow-providers-databricks = "6.9.0"
> apache-airflow-providers-docker = "3.13.0"
> apache-airflow-providers-snowflake = {extras = ["common-compat"], version = "5.7.0"}
> boto3 = "1.35.7"
> pendulum = "^3"
> boto3-stubs = {extras = ["ecs", "resourcegroupstaggingapi"], version = "1.35.7"}
> <custom-internal-package> = {version = "^0.2.43", source = "codeartifact"}
> deepmerge = ">=2.0"
> ```
> 
> Please let me know if there's anything else I can provide. Thank you for all the help!

What are the requirements of the custom-internal-package dependency here? Can you share those as well?


My next most feasible guess is that something in this dependency closure is downgrading the provider version, but I think you would have seen that. Otherwise, the only other difference I can guess is something to do with your AWS networking/compute configuration. So far you're the only user complaining of your particular issue of logs still missing from tail after the fix from @vincbeck.

**Comment by yaningz** [comment]
Hi @o-nikolas 

> Are you using MWAA version 2.10.1?

Yes, I verified the version of the environments in the AWS console just now.

> What are the requirements of the custom-internal-package dependency here? Can you share those as well?

The full package dependency is: `airflow-provider-om1` (our internal Airflow provider code) -> internal package CLI -> internal package.

`airflow-provider-om1`
```
[tool.poetry.dependencies]
python = ">=3.11,<3.12"
apache-airflow = "2.10.1"
apache-airflow-providers-amazon = "8.28.0"
apache-airflow-providers-databricks = "6.9.0"
apache-airflow-providers-docker = "3.13.0"
apache-airflow-providers-snowflake = {extras = ["common-compat"], version = "5.7.0"}
boto3 = "1.35.7"
pendulum = "^3"
boto3-stubs = {extras = ["ecs", "resourcegroupstaggingapi"], version = "1.35.7"}
<internal-package-cli> = {version = "^0.2.43", source = "codeartifact"}
deepmerge = ">=2.0"
```

Internal package CLI
```
[tool.poetry.dependencies]
boto3 = "^1.34"
boto3-stubs = {extras = ["essential"], version = "^1.34"}
cement = "^3.0"
certifi = "^2024.07.04"
colorlog = "^6.7.0"
#docker = "^6.0.1"
jinja2 = "^3.1.2"
jsonschema = "^4.17.1"
<internal-package> = {path = "../parser"}
pyhocon = "^0.3.59"
python = "^3.11"
python-terraform = "^0.10.1"
pyyaml = "^6.0.0"
urllib3 = "^2.0"
```

Internal package (parser)
```
[tool.poetry.dependencies]
boto3 = "^1.34"
boto3-stubs = {extras = ["essential"], version = "^1.34"}
certifi = "^2024.07.04"
jsonschema = "^4.17.1"
pyhocon = "^0.3.59"
python = "^3.9"
```

> Otherwise, the only other difference I can guess is something to do with your AWS networking/compute configuration.

I'm not sure how to debug this, but I will share this issue with our infrastructure engineers who would be more equipped to answer particular questions about potential failure points of our networking configuration.

> Your issue seems to be different from @yaningz, they are missing the tail of their logs, and you are missing the head. I'm still unable to reproduce either case.

Additionally, regarding your question to @mjordyn , Jordyn is a member of a different team at our organization facing a similar but distinct issue (I manage the team tasked with managing the platform and our provider code, Jordyn's team is a user team implementing DAGs). Her issue was discovered in developing a separate task to run within our MWAA environments. All environments are on `apache-airflow = "2.10.1"`.

**Comment by yaningz** [comment]
@o-nikolas Here is the full constraints file.

```

#
# This constraints file was automatically generated on 2024-09-02T09:27:12.793675
# via "eager-upgrade" mechanism of PIP. For the "v2-10-test" branch of Airflow.
# This variant of constraints install uses the HEAD of the branch version for 'apache-airflow' but installs
# the providers from PIP-released packages at the moment of the constraint generation.
#
# Those constraints are actually those that regular users use to install released version of Airflow.
# We also use those constraints after "apache-airflow" is released and the constraints are tagged with
# "constraints-X.Y.Z" tag to build the production image for that version.
#
# This constraints file is meant to be used only in the "apache-airflow" installation command and not
# in all subsequent pip commands. By using a constraints.txt file, we ensure that solely the Airflow
# installation step is reproducible. Subsequent pip commands may install packages that would have
# been incompatible with the constraints used in Airflow reproducible installation step. Finally, pip
# commands that might change the installed version of apache-airflow should include "apache-airflow==X.Y.Z"
# in the list of install targets to prevent Airflow accidental upgrade or downgrade.
#
# Typical installation process of airflow for Python 3.8 is (with random selection of extras and custom
# dependencies added), usually consists of two steps:
#
# 1. Reproducible installation of airflow with selected providers (note constraints are used):
#
# pip install "apache-airflow[celery,cncf.kubernetes,google,amazon,snowflake]==X.Y.Z" \
#     --constraint \
#    "https://raw.githubusercontent.com/apache/airflow/constraints-X.Y.Z/constraints-3.11.txt"
#
# 2. Installing own dependencies that are potentially not matching the constraints (note constraints are not
#    used, and apache-airflow==X.Y.Z is used to make sure there is no accidental airflow upgrade/downgrade.
#
# pip install "apache-airflow==X.Y.Z" "snowflake-connector-python[pandas]=N.M.O"
#
Authlib==1.3.2
ConfigUpdater==3.2
Deprecated==1.2.14
Events==0.5
Flask-AppBuilder==4.5.0
Flask-Babel==2.0.0
Flask-Bcrypt==1.0.1
Flask-Caching==2.3.0
Flask-JWT-Extended==4.6.0
Flask-Limiter==3.8.0
Flask-Login==0.6.3
Flask-SQLAlchemy==2.5.1
Flask-Session==0.5.0
Flask-WTF==1.2.1
Flask==2.2.5
GitPython==3.1.43
JPype1==1.5.0
JayDeBeApi==1.2.3
Jinja2==3.1.4
Js2Py==0.74
Mako==1.3.5
Markdown==3.7
MarkupSafe==2.1.5
PyAthena==3.9.0
PyGithub==2.4.0
PyHive==0.7.0
PyJWT==2.9.0
PyNaCl==1.5.0
PyYAML==6.0.2
Pygments==2.18.0
SQLAlchemy-JSONField==1.0.2
SQLAlchemy-Utils==0.41.2
SQLAlchemy==1.4.53
SecretStorage==3.3.3
Sphinx==5.3.0
WTForms==3.1.2
Werkzeug==2.2.3
adal==1.2.7
adlfs==2024.7.0
aiobotocore==2.14.0
aiofiles==23.2.1
aiohappyeyeballs==2.4.0
aiohttp==3.10.5
aioitertools==0.12.0
aioresponses==0.7.6
aiosignal==1.3.1
airflow-provider-hex==0.1.9
airflow-provider-om1==2.10.0
alabaster==0.7.16
alembic==1.13.2
alibabacloud-adb20211201==2.0.0
alibabacloud-tea==0.3.9
alibabacloud_credentials==0.3.5
alibabacloud_endpoint_util==0.0.3
alibabacloud_gateway_spi==0.0.2
alibabacloud_openapi_util==0.2.2
alibabacloud_tea_openapi==0.3.11
alibabacloud_tea_util==0.3.13
alibabacloud_tea_xml==0.0.2
aliyun-python-sdk-core==2.15.2
aliyun-python-sdk-kms==2.16.5
amqp==5.2.0
analytics-python==1.2.9
annotated-types==0.7.0
ansicolors==1.1.8
anyascii==0.3.2
anyio==4.4.0
apache-airflow-providers-airbyte==3.9.0
apache-airflow-providers-alibaba==2.9.0
apache-airflow-providers-amazon==8.28.0
apache-airflow-providers-apache-beam==5.8.0
apache-airflow-providers-apache-cassandra==3.6.0
apache-airflow-providers-apache-drill==2.8.0
apache-airflow-providers-apache-druid==3.11.0
apache-airflow-providers-apache-flink==1.5.0
apache-airflow-providers-apache-hdfs==4.5.0
apache-airflow-providers-apache-hive==8.2.0
apache-airflow-providers-apache-iceberg==1.1.0
apache-airflow-providers-apache-impala==1.5.0
apache-airflow-providers-apache-kafka==1.6.0
apache-airflow-providers-apache-kylin==3.7.0
apache-airflow-providers-apache-livy==3.9.0
apache-airflow-providers-apache-pig==4.5.0
apache-airflow-providers-apache-pinot==4.5.0
apache-airflow-providers-apache-spark==4.10.0
apache-airflow-providers-apprise==1.4.0
apache-airflow-providers-arangodb==2.6.0
apache-airflow-providers-asana==2.6.0
apache-airflow-providers-atlassian-jira==2.7.0
apache-airflow-providers-celery==3.8.1
apache-airflow-providers-cloudant==3.6.0
apache-airflow-providers-cncf-kubernetes==8.4.1
apache-airflow-providers-cohere==1.3.0
apache-airflow-providers-common-compat==1.2.0
apache-airflow-providers-common-io==1.4.0
apache-airflow-providers-common-sql==1.16.0
apache-airflow-providers-databricks==6.9.0
apache-airflow-providers-datadog==3.7.0
apache-airflow-providers-dbt-cloud==3.10.0
apache-airflow-providers-dingding==3.6.0
apache-airflow-providers-discord==3.8.0
apache-airflow-providers-docker==3.13.0
apache-airflow-providers-elasticsearch==5.5.0
apache-airflow-providers-exasol==4.6.0
apache-airflow-providers-fab==1.3.0
apache-airflow-providers-facebook==3.6.0
apache-airflow-providers-ftp==3.11.0
apache-airflow-providers-github==2.7.0
apache-airflow-providers-google==10.22.0
apache-airflow-providers-grpc==3.6.0
apache-airflow-providers-hashicorp==3.8.0
apache-airflow-providers-http==4.13.0
apache-airflow-providers-imap==3.7.0
apache-airflow-providers-influxdb==2.7.0
apache-airflow-providers-jdbc==4.5.0
apache-airflow-providers-jenkins==3.7.0
apache-airflow-providers-microsoft-azure==10.4.0
apache-airflow-providers-microsoft-mssql==3.9.0
apache-airflow-providers-microsoft-psrp==2.8.0
apache-airflow-providers-microsoft-winrm==3.6.0
apache-airflow-providers-mongo==4.2.0
apache-airflow-providers-mysql==5.7.0
apache-airflow-providers-neo4j==3.7.0
apache-airflow-providers-odbc==4.7.0
apache-airflow-providers-openai==1.3.0
apache-airflow-providers-openfaas==3.6.0
apache-airflow-providers-openlineage==1.11.0
apache-airflow-providers-opensearch==1.4.0
apache-airflow-providers-opsgenie==5.7.0
apache-airflow-providers-oracle==3.11.0
apache-airflow-providers-pagerduty==3.8.0
apache-airflow-providers-papermill==3.8.0
apache-airflow-providers-pgvector==1.3.0
apache-airflow-providers-pinecone==2.1.0
apache-airflow-providers-postgres==5.12.0
apache-airflow-providers-presto==5.6.0
apache-airflow-providers-qdrant==1.2.0
apache-airflow-providers-redis==3.8.0
apache-airflow-providers-salesforce==5.8.0
apache-airflow-providers-samba==4.8.0
apache-airflow-providers-segment==3.6.0
apache-airflow-providers-sendgrid==3.6.0
apache-airflow-providers-sftp==4.11.0
apache-airflow-providers-singularity==3.6.0
apache-airflow-providers-slack==8.9.0
apache-airflow-providers-smtp==1.8.0
apache-airflow-providers-snowflake==5.7.0
apache-airflow-providers-sqlite==3.9.0
apache-airflow-providers-ssh==3.13.1
apache-airflow-providers-tableau==4.6.0
apache-airflow-providers-tabular==1.6.1
apache-airflow-providers-telegram==4.6.0
apache-airflow-providers-teradata==2.6.0
apache-airflow-providers-trino==5.8.0
apache-airflow-providers-vertica==3.9.0
apache-airflow-providers-weaviate==2.1.0
apache-airflow-providers-yandex==3.12.0
apache-airflow-providers-ydb==1.3.0
apache-airflow-providers-zendesk==4.8.0
apache-beam==2.56.0
apispec==6.6.1
apprise==1.8.1
argcomplete==3.5.0
asana==3.2.3
asgiref==3.8.1
asn1crypto==1.5.1
astroid==2.15.8
asttokens==2.4.1
asyncssh==2.16.0
atlasclient==1.0.0
atlassian-python-api==3.41.15
attrs==24.2.0
aws-sam-translator==1.91.0
aws-xray-sdk==2.14.0
azure-batch==14.2.0
azure-common==1.1.28
azure-core==1.30.2
azure-cosmos==4.7.0
azure-datalake-store==0.0.53
azure-identity==1.17.1
azure-keyvault-secrets==4.8.0
azure-kusto-data==4.5.1
azure-mgmt-containerinstance==10.1.0
azure-mgmt-containerregistry==10.3.0
azure-mgmt-core==1.4.0
azure-mgmt-cosmosdb==9.5.1
azure-mgmt-datafactory==9.0.0
azure-mgmt-datalake-nspkg==3.0.1
azure-mgmt-datalake-store==0.5.0
azure-mgmt-nspkg==3.0.2
azure-mgmt-resource==23.1.1
azure-mgmt-storage==21.2.1
azure-nspkg==3.0.2
azure-servicebus==7.12.2
azure-storage-blob==12.22.0
azure-storage-file-datalake==12.16.0
azure-storage-file-share==12.17.0
azure-synapse-artifacts==0.19.0
azure-synapse-spark==0.7.0
babel==2.16.0
backoff==2.2.1
backports.tarfile==1.2.0
bcrypt==4.2.0
beautifulsoup4==4.12.3
billiard==4.2.0
bitarray==2.9.2
black==24.8.0
blinker==1.8.2
boto3==1.35.7
botocore==1.35.7
cachelib==0.9.0
cachetools==5.5.0
cassandra-driver==3.29.1
cattrs==24.1.0
celery==5.4.0
certifi==2024.8.30
cffi==1.17.0
cfgv==3.4.0
cfn-lint==1.11.1
cgroupspy==0.2.3
chardet==5.2.0
charset-normalizer==3.3.2
checksumdir==1.2.0
ciso8601==2.3.1
click-didyoumean==0.3.1
click-plugins==1.1.1
click-repl==0.3.0
click==8.1.7
clickclick==20.10.2
cloudant==2.15.0
cloudpickle==2.2.1
cohere==4.57
colorama==0.4.6
colorlog==6.8.2
comm==0.2.2
confluent-kafka==2.5.0
connexion==2.14.2
coverage==7.6.1
crcmod==1.7
cron-descriptor==1.4.5
croniter==3.0.3
cryptography==42.0.8
curlify==2.2.1
databricks-sql-connector==2.9.6
datadog==0.50.0
db-dtypes==1.3.0
debugpy==1.8.5
decorator==5.1.1
defusedxml==0.7.1
deltalake==0.18.2
diagrams==0.23.4
dill==0.3.1.1
distlib==0.3.8
distro==1.9.0
dnspython==2.6.1
docker==7.1.0
docopt==0.6.2
docstring_parser==0.16
docutils==0.16
duckdb==1.0.0
elastic-transport==8.15.0
elasticsearch==8.15.0
email_validator==2.2.0
entrypoints==0.4
eralchemy2==1.4.1
et-xmlfile==1.1.0
eventlet==0.36.1
execnet==2.1.1
executing==2.1.0
facebook_business==20.0.2
fastavro==1.9.5
fasteners==0.19
fastjsonschema==2.20.0
filelock==3.15.4
flower==2.0.1
frozenlist==1.4.1
fsspec==2024.6.1
future==1.0.0
gcloud-aio-auth==5.3.2
gcloud-aio-bigquery==7.1.0
gcloud-aio-storage==9.3.0
gcsfs==2024.6.1
geomet==0.2.1.post1
gevent==24.2.1
gitdb==4.0.11
google-ads==25.0.0
google-analytics-admin==0.23.0
google-api-core==2.19.2
google-api-python-client==2.143.0
google-auth-httplib2==0.2.0
google-auth-oauthlib==1.2.1
google-auth==2.34.0
google-cloud-aiplatform==1.64.0
google-cloud-appengine-logging==1.4.5
google-cloud-audit-log==0.3.0
google-cloud-automl==2.13.5
google-cloud-batch==0.17.26
google-cloud-bigquery-datatransfer==3.15.6
google-cloud-bigquery==3.20.1
google-cloud-bigtable==2.26.0
google-cloud-build==3.24.2
google-cloud-compute==1.19.2
google-cloud-container==2.50.0
google-cloud-core==2.4.1
google-cloud-datacatalog==3.20.1
google-cloud-dataflow-client==0.8.12
google-cloud-dataform==0.5.11
google-cloud-dataplex==2.2.2
google-cloud-dataproc-metastore==1.15.5
google-cloud-dataproc==5.10.2
google-cloud-dlp==3.22.0
google-cloud-kms==2.24.2
google-cloud-language==2.14.0
google-cloud-logging==3.11.2
google-cloud-memcache==1.9.5
google-cloud-monitoring==2.22.2
google-cloud-orchestration-airflow==1.13.1
google-cloud-os-login==2.14.6
google-cloud-pubsub==2.23.0
google-cloud-redis==2.15.5
google-cloud-resource-manager==1.12.5
google-cloud-run==0.10.8
google-cloud-secret-manager==2.20.2
google-cloud-spanner==3.48.0
google-cloud-speech==2.27.0
google-cloud-storage-transfer==1.12.0
google-cloud-storage==2.18.2
google-cloud-tasks==2.16.5
google-cloud-texttospeech==2.17.2
google-cloud-translate==3.16.0
google-cloud-videointelligence==2.13.5
google-cloud-vision==3.7.4
google-cloud-workflows==1.14.5
google-crc32c==1.5.0
google-re2==1.1.20240702
google-resumable-media==2.7.2
googleapis-common-protos==1.65.0
graphql-core==3.2.3
graphviz==0.20.3
greenlet==3.0.3
grpc-google-iam-v1==0.13.1
grpc-interceptor==0.15.4
grpcio-gcp==0.2.2
grpcio-health-checking==1.62.3
grpcio-status==1.62.3
grpcio-tools==1.62.3
grpcio==1.66.1
gssapi==1.8.3
gunicorn==23.0.0
h11==0.14.0
h2==4.1.0
hatch==1.12.0
hatchling==1.25.0
hdfs==2.7.3
hmsclient==0.1.1
hpack==4.0.0
httpcore==1.0.5
httplib2==0.22.0
httpx==0.27.0
humanize==4.10.0
hvac==2.3.0
hyperframe==6.0.1
hyperlink==21.0.0
icdiff==2.0.7
identify==2.6.0
idna==3.8
ijson==3.3.0
imagesize==1.4.1
importlib-metadata==6.11.0
importlib_resources==6.4.4
impyla==0.19.0
incremental==24.7.2
inflection==0.5.1
influxdb-client==1.45.0
iniconfig==2.0.0
ipdb==0.13.13
ipykernel==6.29.5
ipython==8.27.0
isodate==0.6.1
itsdangerous==2.2.0
jaraco.classes==3.4.0
jaraco.context==6.0.1
jaraco.functools==4.0.2
jedi==0.19.1
jeepney==0.8.0
jiter==0.5.0
jmespath==0.10.0
joserfc==1.0.0
json-merge-patch==0.2
jsondiff==2.2.1
jsonpatch==1.33
jsonpath-ng==1.6.1
jsonpickle==3.3.0
jsonpointer==3.0.0
jsonschema-path==0.3.3
jsonschema-specifications==2023.12.1
jsonschema==4.23.0
jupyter_client==8.6.2
jupyter_core==5.7.2
keyring==25.3.0
kombu==5.4.0
krb5==0.6.0
kubernetes==30.1.0
kubernetes_asyncio==30.1.0
kylinpy==2.8.4
lazy-object-proxy==1.10.0
ldap3==2.9.1
limits==3.13.0
linkify-it-py==2.0.3
lockfile==0.12.2
looker-sdk==24.14.0
lxml==5.3.0
lz4==4.3.3
markdown-it-py==3.0.0
marshmallow-oneofschema==3.1.1
marshmallow-sqlalchemy==0.28.2
marshmallow==3.22.0
matplotlib-inline==0.1.7
mdit-py-plugins==0.4.1
mdurl==0.1.2
mergedeep==1.3.4
methodtools==0.4.7
microsoft-kiota-abstractions==1.3.3
microsoft-kiota-authentication-azure==1.1.0
microsoft-kiota-http==1.3.3
mmh3==4.1.0
mongomock==4.1.2
more-itertools==10.4.0
moto==5.0.11
mpmath==1.3.0
msal-extensions==1.2.0
msal==1.30.0
msgraph-core==1.1.2
msrest==0.7.1
msrestazure==0.6.4.post1
multi_key_dict==2.0.3
multidict==6.0.5
mypy-boto3-appflow==1.35.0
mypy-boto3-rds==1.35.0
mypy-boto3-redshift-data==1.35.10
mypy-boto3-s3==1.35.2
mypy-extensions==1.0.0
mypy==1.9.0
mysql-connector-python==9.0.0
mysqlclient==2.2.4
nbclient==0.10.0
nbformat==5.10.4
neo4j==5.24.0
nest-asyncio==1.6.0
networkx==3.3
nh3==0.2.18
nodeenv==1.9.1
numpy==1.26.4
oauthlib==3.2.2
objsize==0.7.0
openai==1.43.0
openapi-schema-validator==0.6.2
openapi-spec-validator==0.7.1
opencv-python==4.10.0.84
openlineage-integration-common==1.21.1
openlineage-python==1.21.1
openlineage_sql==1.21.1
openpyxl==3.1.5
opensearch-py==2.7.1
opentelemetry-api==1.27.0
opentelemetry-exporter-otlp-proto-common==1.27.0
opentelemetry-exporter-otlp-proto-grpc==1.27.0
opentelemetry-exporter-otlp-proto-http==1.27.0
opentelemetry-exporter-otlp==1.27.0
opentelemetry-exporter-prometheus==0.48b0
opentelemetry-proto==1.27.0
opentelemetry-sdk==1.27.0
opentelemetry-semantic-conventions==0.48b0
opsgenie-sdk==2.1.5
oracledb==2.4.1
ordered-set==4.1.0
orjson==3.10.7
oss2==2.19.0
packaging==24.1
pandas-gbq==0.23.1
pandas-stubs==2.2.2.240807
pandas==2.1.4
papermill==2.6.0
paramiko==3.4.1
parso==0.8.4
pathable==0.4.3
pathspec==0.12.1
pbr==6.1.0
pdpyras==5.2.0
pendulum==3.0.0
pexpect==4.9.0
pgvector==0.3.2
pillow==11.0.0
pinecone-client==5.0.1
pinecone-plugin-inference==1.0.3
pinecone-plugin-interface==0.0.7
pinotdb==5.6.0
pipdeptree==2.23.1
pipx==1.7.1
pkginfo==1.10.0
platformdirs==4.2.2
pluggy==1.5.0
ply==3.11
plyvel==1.5.1
portalocker==2.10.1
pprintpp==0.4.0
pre-commit==3.8.0
presto-python-client==0.8.4
prison==0.2.1
prometheus_client==0.20.0
prompt_toolkit==3.0.47
proto-plus==1.24.0
protobuf==4.25.4
psutil==6.0.0
psycopg2-binary==2.9.9
ptyprocess==0.7.0
pure-sasl==0.6.2
pure_eval==0.2.3
py-partiql-parser==0.5.5
py4j==0.10.9.7
pyOpenSSL==24.2.1
pyarrow-hotfix==0.6
pyarrow==14.0.2
pyasn1==0.6.0
pyasn1_modules==0.4.0
pycountry==24.6.1
pycparser==2.22
pycryptodome==3.20.0
pydantic==2.8.2
pydantic_core==2.20.1
pydata-google-auth==1.8.2
pydot==1.4.2
pydruid==0.6.9
pyenchant==3.2.2
pyexasol==0.26.0
pygraphviz==1.13
pyiceberg==0.7.1
pyjsparser==2.7.1
pykerberos==1.2.4
pymongo==4.8.0
pymssql==2.3.1
pyodbc==5.1.0
pyparsing==3.1.4
pypsrp==0.8.1
pyspark==3.5.2
pyspnego==0.11.1
pytest-asyncio==0.24.0
pytest-cov==5.0.0
pytest-custom-exit-code==0.3.0
pytest-icdiff==0.9
pytest-instafail==0.5.0
pytest-mock==3.14.0
pytest-rerunfailures==14.0
pytest-timeouts==1.2.1
pytest-xdist==3.6.1
pytest==8.3.2
python-arango==8.1.0
python-daemon==3.0.1
python-dateutil==2.9.0.post0
python-dotenv==1.0.1
python-http-client==3.3.7
python-jenkins==1.8.2
python-ldap==3.4.4
python-nvd3==0.16.0
python-slugify==8.0.4
python-telegram-bot==21.5
python3-saml==1.16.0
pytz==2024.1
pywinrm==0.5.0
pyzmq==26.2.0
qdrant-client==1.11.1
reactivex==4.0.4
readme_renderer==43.0
redis==5.0.8
redshift-connector==2.1.3
referencing==0.35.1
regex==2024.7.24
requests-file==2.1.0
requests-kerberos==0.15.0
requests-mock==1.12.1
requests-oauthlib==1.3.1
requests-toolbelt==1.0.0
requests==2.32.3
requests_ntlm==1.3.0
responses==0.25.3
restructuredtext_lint==1.4.0
rfc3339-validator==0.1.4
rfc3986==2.0.0
rich-argparse==1.5.2
rich-click==1.8.3
rich==13.8.0
rpds-py==0.20.0
rsa==4.9
ruff==0.5.5
s3fs==2024.6.1
s3transfer==0.10.2
scramp==1.4.5
scrapbook==0.5.0
semver==3.0.2
sendgrid==6.11.0
sentinels==1.0.0
sentry-sdk==2.13.0
setproctitle==1.3.3
shapely==2.0.6
shellingham==1.5.4
simple-salesforce==1.12.6
six==1.16.0
slack_sdk==3.31.0
smbprotocol==1.14.0
smmap==5.0.1
sniffio==1.3.1
snowballstemmer==2.2.0
snowflake-connector-python==3.12.1
snowflake-sqlalchemy==1.6.1
sortedcontainers==2.4.0
soupsieve==2.6
sphinx-airflow-theme==0.0.12
sphinx-argparse==0.5.0
sphinx-autoapi==2.1.1
sphinx-copybutton==0.5.2
sphinx-jinja==2.0.2
sphinx-rtd-theme==2.0.0
sphinx_design==0.6.0
sphinxcontrib-applehelp==2.0.0
sphinxcontrib-devhelp==2.0.0
sphinxcontrib-htmlhelp==2.1.0
sphinxcontrib-httpdomain==1.8.1
sphinxcontrib-jquery==4.1
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==2.0.0
sphinxcontrib-redoc==1.6.0
sphinxcontrib-serializinghtml==1.1.5
sphinxcontrib-spelling==8.0.0
spython==0.3.13
sqlalchemy-bigquery==1.11.0
sqlalchemy-redshift==0.8.14
sqlalchemy-spanner==1.7.0
sqlalchemy_drill==1.1.5
sqlparse==0.5.1
sshtunnel==0.4.0
stack-data==0.6.3
starkbank-ecdsa==2.2.0
statsd==4.0.1
std-uritemplate==1.0.6
strictyaml==1.7.3
sympy==1.13.2
tableauserverclient==0.32
tabulate==0.9.0
tenacity==8.5.0
teradatasql==20.0.0.15
teradatasqlalchemy==20.0.0.1
termcolor==2.4.0
text-unidecode==1.3
thrift-sasl==0.4.3
thrift==0.16.0
time-machine==2.15.0
tomli_w==1.0.0
tomlkit==0.13.2
tornado==6.4.1
towncrier==24.8.0
tqdm==4.66.5
traitlets==5.14.3
trino==0.329.0
trove-classifiers==2024.7.2
twine==5.1.1
typed-ast==1.5.5
types-Deprecated==1.2.9.20240311
types-Markdown==3.7.0.20240822
types-PyMySQL==1.1.0.20240524
types-PyYAML==6.0.12.20240808
types-aiofiles==24.1.0.20240626
types-certifi==2021.10.8.3
types-cffi==1.16.0.20240331
types-croniter==3.0.3.20240731
types-docutils==0.21.0.20240724
types-paramiko==3.4.0.20240423
types-protobuf==5.27.0.20240626
types-pyOpenSSL==24.1.0.20240722
types-python-dateutil==2.9.0.20240821
types-python-slugify==8.0.2.20240310
types-pytz==2024.1.0.20240417
types-redis==4.6.0.20240819
types-requests==2.32.0.20240712
types-setuptools==74.0.0.20240831
types-tabulate==0.9.0.20240106
types-termcolor==1.1.6.2
types-toml==0.10.8.20240310
typing_extensions==4.12.2
tzdata==2024.1
tzlocal==5.2
uc-micro-py==1.0.3
unicodecsv==0.14.1
universal_pathlib==0.2.3
uritemplate==4.1.1
urllib3==2.2.2
userpath==1.9.2
uv==0.4.1
validators==0.33.0
vertica-python==1.4.0
vine==5.1.0
virtualenv==20.26.3
watchtower==3.3.1
wcwidth==0.2.13
weaviate-client==4.7.1
websocket-client==1.8.0
wirerope==0.4.7
wrapt==1.16.0
xmlsec==1.3.14
xmltodict==0.13.0
yamllint==1.35.1
yandex-query-client==0.1.4
yandexcloud==0.315.0
yarl==1.9.7
ydb==3.16.0
zeep==4.2.1
zenpy==2.0.50
zipp==3.20.1
zope.event==5.0
zope.interface==7.0.3
zstandard==0.23.0
```

**Comment by jreslock** [comment]
> I've tried to reproduce the issue today, running ECS tasks that log 5, 50, 500, and 5000 lines of logging. Trying to run each one multiple times, and every time I got every log line reliably. So there is some piece missing either in my repro or the setup you all have running. This actually isn't too surprising, since if it was very easy to reproduce, we'd have many people cutting tickets (ECS is one of the most used AWS operators by far).
> 
> Is there anything in your infra setup that would cause CloudWatch to take longer to become consistent? Are you logging across regions? Or connecting through VCPE or some other gateway that would slow down network traffic out of your VPC? Any more specific you can provide will be very helpful to try reproduce something!

I work with @yaningz and can speak a bit about our network configuration.

TL;DR this is a single VPC with 3 private subnets, a VPC endpoint and 2 security groups.

-  Airflow (MWAA) running in a single VPC attached to two private subnets and a single security group. 
-  ECS tasks launched by the ECSRunTaskOperator are co-located on these same 2 private subnets plus one additional. Each of our VPCs always has 3 private subnets for tasks for AZ redundancy. 
 - Each of the private subnets routes through a NAT Gateway for internet egress
- There is a VPC Endpoint configured for CloudWatch access. This endpoint is associated with all 3 private subnets and has 2 security groups associated, one allowing HTTP/443 ingress from the IPv4 VPC CIDR and the second allowing ingress from itself (any other entity associated with this SG will be allowed)

There is no transit gateway, VPC lattice, cross-account or VPC traversal in play and I will add that all other CloudWatch logging functionality appears to be working as expected.

Could these missing log entries be caused by some unhandled exception in the interaction between `task_log_fetcher`, `waiter`, and `waiter_with_logging`? I tried to walk myself through this code last night but I quickly lost my way.

**Comment by o-nikolas** [comment]
Resolving for now since we cannot repro the issue. If more folks run into the issue and have more datapoints/context to provide we can re-open.
